composer/loggers/mosaicml_logger.py

```diff
@@ -0,0 +1,164 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+from __future__ import annotations
+
+import logging
+import os
+import sys
+import warnings
+from queue import Queue
+from threading import Thread
+from typing import Optional
+
+import requests
+
+from composer.core.logging import LogLevel, RankZeroLoggerBackend, TLogData
+from composer.core.types import JSON, Logger, State, StateDict
+
+_MOSAICML_API_KEY_ENV = "MOSAICML_LOGGER_API_KEY"
+_MOSAICML_LOGGER_URL = "https://api.mosaicml.com/v0/log/metric"
+_JOB_ID_ENV = "MOSAICML_LOGGER_JOB_ID"
+
+_STOP_LOGGING_SIGNAL = "STOP_LOGGING"
+
+log = logging.getLogger(__name__)
+
+
+def _send_data(job_id: str, data: JSON):
+    try:
+        response = requests.post(_MOSAICML_LOGGER_URL,
+                                 headers={"X-MosaicML-API-key": os.environ.get(_MOSAICML_API_KEY_ENV, "")},
+                                 json={
+                                     "experimentID": job_id,
+                                     "data": data,
+                                 },
+                                 timeout=5)
+        if response.status_code >= 400:
+            # Terminate the process in the case of a non-transient error
+            log.error("Posting data to MosaicML backend failed with response code "
+                      f"{response.status_code} and message {response.text}.")
+            sys.exit(1)
+    except requests.exceptions.Timeout as e:
+        # Warn the user in the case of a timeout but don't retry for now
+        log.warning(f"MosaicML logger timed out with error {e}. Logs were not sent to the backend.")
+
+
+class MosaicMLLoggerBackend(RankZeroLoggerBackend):
+    """Log to the MosaicML backend.
+
+    Args:
+        job_id (str, optional): The id of the job to write logs for.
+        creds_file (str, optional): A file containing the MosaicML api_key. If not provided
+            will default to the environment variable MOSAIC_API_KEY.
+        flush_every_n_batches (int): Flush the log data buffer every n batches. (default: ``100``)
+        max_logs_in_buffer (int): The maximum number of log entries allowed in the buffer before a
+            forced flush. (default: ``1000``)
+    """
+
+    def __init__(self,
+                 job_id: Optional[str] = None,
```

Comment:
It is called experiment id other places - should we align on a single term, or use "jobs" for benchmarks and "experiments" for the rest?